packages/userstorage/src/browser/user-storage-service-filesystem.ts

```diff
@@ -18,34 +18,35 @@ import { DisposableCollection, ILogger, Emitter, Event } from '@theia/core/lib/c
 import { UserStorageChangeEvent, UserStorageService } from './user-storage-service';
 import { injectable, inject } from 'inversify';
 import { FileSystemWatcher, FileChangeEvent } from '@theia/filesystem/lib/browser/filesystem-watcher';
+import { EnvVariablesServer } from '@theia/core/lib/common/env-variables';
 import { FileSystem } from '@theia/filesystem/lib/common';
 import URI from '@theia/core/lib/common/uri';
 import { UserStorageUri } from './user-storage-uri';
 
-export const THEIA_USER_STORAGE_FOLDER = '.theia';
-
 @injectable()
 export class UserStorageServiceFilesystemImpl implements UserStorageService {
 
     protected readonly toDispose = new DisposableCollection();
     protected readonly onUserStorageChangedEmitter = new Emitter<UserStorageChangeEvent>();
-    protected readonly userStorageFolder: Promise<URI | undefined>;
+    protected userStorageFolder: Promise<URI | undefined>;
 
     constructor(
         @inject(FileSystem) protected readonly fileSystem: FileSystem,
         @inject(FileSystemWatcher) protected readonly watcher: FileSystemWatcher,
-        @inject(ILogger) protected readonly logger: ILogger
-
+        @inject(ILogger) protected readonly logger: ILogger,
+        @inject(EnvVariablesServer) protected readonly envServer: EnvVariablesServer
     ) {
-        this.userStorageFolder = this.fileSystem.getCurrentUserHome().then(home => {
-            if (home) {
-                const userStorageFolderUri = new URI(home.uri).resolve(THEIA_USER_STORAGE_FOLDER);
-                watcher.watchFileChanges(userStorageFolderUri).then(disposable =>
-                    this.toDispose.push(disposable)
-                );
-                this.toDispose.push(this.watcher.onFilesChanged(changes => this.onDidFilesChanged(changes)));
-                return new URI(home.uri).resolve(THEIA_USER_STORAGE_FOLDER);
-            }
+        this.envServer.getDataFolderName().then(userStorageFolder => {
+            this.userStorageFolder = this.fileSystem.getCurrentUserHome().then(home => {
+                if (home) {
+                    const userStorageFolderUri = new URI(home.uri).resolve(userStorageFolder);
+                    watcher.watchFileChanges(userStorageFolderUri).then(disposable =>
```

Comment:
`userStorageFolderUri` === `new URI('file://' + (await this.envServer.getUserDataFolderPath()))` ?

Reply:
Yes, possible. But just don't want to mix await with promises.
By the way, do you know why another instance of uri is created in return statement? Cannot we just return `userStorageFolderUri`?

Reply:
I meant that the semantic how `getUserDataFolderPath` is constructed can be different from using `Filesystem.getCurrentUserHome`. We should not reimplement it on the client.